consolidate platform readme (#798)
Azure/azure-sdk-for-c: sdk/platform/README.md

You can also implement your own HTTP adapter and use it. This option is in case you want to use an HTTP stack that is not libcurl. Follow the instructions on [use your own http stack implementation](https://github.com/Azure/azure-sdk-for-c#using-your-own-http-stack-implementation).

## Platform

Azure SDK Core depends on some system-specific functions. This functions are not part of the C99 standard library and its implementation depends on system architecture (like a clock, thead sleep or interlock).

Azure SDK provides three platform implementations for you, one for Windows (`az_win32`), another for Linux and MacOS (`az_posix`) and an empty implementation (`az_noplatform`).

See [compiler options](https://github.com/Azure/azure-sdk-for-c#compiler-options) to learn about how to build and use each of this available targets or how to use your own platform implementation.

>Note: `az_noplatform` can be used to link your application and enable Azure Core to run. However, this is not recommended for a production application. It is suggested to be used for testing or Demo.

> Azure SDK for Embedded C is licensed under the [MIT][azure_sdk_for_c_license] license.

Reviewer: ```suggestion
## Contributing

If you'd like to contribute to this library, please read the [contributing guide][azure_sdk_for_c_contributing] to learn more about how to build and test the code.

### License

Azure SDK for Embedded C is licensed under the [MIT][azure_sdk_for_c_license] license.

<!-- LINKS -->
[azure_sdk_for_c_contributing]: https://github.com/Azure/azure-sdk-for-c/blob/master/CONTRIBUTING.md

```

# Azure SDK Platform

> The Azure SDK platform provides two abstractions with some built-in implementations that the caller can use or override with their own behavior. The first one is a transport adapter with helpers to allow clients to communicate with Azure. The second one is operating system specific implementations for the small set of features  that the Azure SDK needs that are not available as part of the C99 standard library (such as thread sleep).

Reviewer 2: nit: Remove extra whitespace
```suggestion
The Azure SDK platform provides two abstractions with some built-in implementations that the caller can use or override with their own behavior. The first one is a transport adapter with helpers to allow clients to communicate with Azure. The second one is operating system specific implementations for the small set of features that the Azure SDK needs that are not available as part of the C99 standard library (such as thread sleep).
```

## HTTP Transport Adapter

> The Azure SDK has its own HTTP request and response structures. For this reason, an HTTP transport adapter is required to be implemented to use the Azure SDK HTTP request and response as input for any HTTP stack (like libcurl or win32). The library allows clients to communicate with Azure.

Reviewer 2: > for any HTTP stack (like libcurl or win32)

What does `win32` mean here?

This wasn't addressed earlier:
> Did you mean `win32` or something else, like `WinINet`?

Author: well, we call it win32.  Honestly I am not really sure about the proper way to call the Windows HTTP stack.
Making a quick search, my understanding is that WinINet is part of win32.
And there's also WinHTTP within win32 API as a way to use HTTP stack.
https://docs.microsoft.com/en-us/windows/win32/wininet/about-wininet

So, I believe we are OK saying win32 as a general Windows API to be used for HTTP.  What do you think?

Reviewer 2: > So, I believe we are OK saying win32 as a general Windows API to be used for HTTP. What do you think?

Sounds good Thanks for clarifying.

Azure SDK provides one implementation for libcurl (`az_curl`). To consume this implementation, link your application against `az_core` and `az_curl` (if using Cmake) and then HTTP requests will be sent using `libcurl`.

>Note: See [Compiler Options](https://github.com/Azure/azure-sdk-for-c#compiler-options). You have to turn on building curl transport in order to have this adapter available.

> The Azure SDK will also build and provide empty HTTP adapter stubs called `az_nohttp`. This target allows you to build `az_core` without any specific HTTP adapter. Use this option when you won't use any HTTP specific APIs from the Azure SDK.

Reviewer 2: ```suggestion
The Azure SDK also provides empty HTTP adapter stubs called `az_nohttp`. This target allows you to build `az_core` without any specific HTTP adapter. Use this option when you won't use any HTTP specific APIs from the Azure SDK.
```

## Platform

Azure SDK Core depends on some system-specific functions. These functions are not part of the C99 standard library and their implementation depends on system architecture (for example a clock, thread sleep, or interlock).

Azure SDK provides three platform implementations for you, one for Windows (`az_win32`), another for Linux and MacOS (`az_posix`) and an empty implementation (`az_noplatform`).

See [compiler options](https://github.com/Azure/azure-sdk-for-c#compiler-options) to learn about how to build and use each of these available targets or how to use your own platform implementation.

>Note: `az_noplatform` can be used to link your application and enable Azure Core to run. However, this is not recommended for a production application. It is suggested to be used for testing or just for getting started.

## Contributing

If you'd like to contribute to this library, please read the [contributing guide][azure_sdk_for_c_contributing] to learn more about how to build and test the code.

### License

> Azure SDK for Embedded C is licensed under the [MIT][azure_sdk_for_c_license] license.

Reviewer 2: The `[azure_sdk_for_c_license]` link is missing at the bottom.
![image](https://user-images.githubusercontent.com/6527137/83709214-e7dda800-a5d2-11ea-9b6e-3a719d5b87c2.png)

Author: fixed

# Azure SDK Platform

Azure SDK Platform contains two important components. The first component is related to HTTP and the second one is implementation for system specific features that Azure Core uses (like thread sleep).

> ## HTTP Adapter

Reviewer 2: ```suggestion
## HTTP Transport Adapter
```

> Azure SDK has its own HTTP Request and Response structures. For this reason, an HTTP adapter is required to be implemented to use Azure SDK HTTP Request and Response as input for any HTTP Stack library (like libcurl or win32).

Reviewer 2: Is what we are saying here true? Do we really have a transport policy and abstractions **because we have our own HTTP request and response structure**? That's what this paragraph reads like but I am not sure if that's the reason why we have this transport layer.

> Azure SDK has its own HTTP Request and Response structures. For this reason, an HTTP adapter is required to be implemented to use Azure SDK HTTP Request and Response as input for any HTTP Stack library (like libcurl or win32).

Author: yeah, that is the reason.  C99 does not offer a default HTTP request and HTTP stack to send/receive requests.
Our SDK brings an specific design for an HTTP Request and Response. And due to that, we need adapters for them to be used with any HTTP stack like libcurl.

We could use libcurl directly, and create our requests on a libcurl-ready way, but that would make us libcurl-only.  Having our on HTTP request makes be independent from any HTTP stack (but requires to build adapters for any of them)
